### ISSUE TITLE
Poor quality of image scaling in Viewer

### ISSUE THREAD

**Issue by corvus1** [doublecmd/doublecmd#680]
![Screenshot_20221125_082634](https://user-images.githubusercontent.com/4068104/203901388-b4d5975b-5cb7-406a-8e9e-a35fc090aa94.png)

![Screenshot_20221125_082751](https://user-images.githubusercontent.com/4068104/203901415-c2c344cd-a8f9-4c1e-a43d-cab8924e7e62.png)

Whatever is doing the scaling for Viewer is really not doing a great job.

**Comment by Skif-off** [comment]
Have you checked the same scaling of this file in other programs?

P.S. Once I tried to figure out what scaling method Lazarus uses and whether it can be changed, but I could not :( Maybe just "nearest neighbor" - poor quality when reduced in size, but very easy/simple and very fast.

**Comment by corvus1** [comment]
![Screenshot_20221125_193358](https://user-images.githubusercontent.com/4068104/204017076-491c4034-3ac2-4526-b7dc-1a9dcedfa3aa.png)

The same image, scaled to the same size in gwenview.
I can at least read the text. 
I get that it's fast, but the quality in Viewer I think is just piss-poor.

**Comment by Skif-off** [comment]
[src/uresample.pas](https://github.com/doublecmd/doublecmd/blob/master/src/uresample.pas)

As I understand, DC uses Hermite for thumbnails, but I did not understand what is used for scaling in the viewer (I think, "Box" aka "Nearest Neighbour").
There is no perfect choice: it depends on the size, content, increase/decrease (+ coefficient). For example, GIMP uses Bicubic (is Mitchell the same as Bicubic?), but ImageMagic uses Mitchell (if new size > original size) or Lanczos (other cases) - two parameters?

**Comment by corvus1** [comment]
I looked at the Box filter, that smells like the '80s, I'm sure it's very fast, but I don't think it can be used in 2020s. It's caveman.

https://github.com/doublecmd/doublecmd/blob/52a9d37f9cca16505b38c6b9b38c0640918d1ce7/src/uresample.pas#L155-L165

I'm sure we can do better, without a noticeable performance penalty.

**Comment by corvus1** [comment]
A very quick look around tells me that Lazarus already has something a bit better https://wiki.freepascal.org/BGRABitmap_tutorial_13#Canvas_and_BGRACanvas

**Comment by Skif-off** [comment]
> I looked at the Box filter, that smells like the '80s, ...  I don't think it can be used in 2020s. It's caveman.

We use `0` over 2000 years, looks like mammoth shit :)

> A very quick look around tells me that Lazarus already has something a bit better https://wiki.freepascal.org/BGRABitmap_tutorial_13#Canvas_and_BGRACanvas

But example with... `rmSimpleStretch` :)))

> I'm sure we can do better, without a noticeable performance penalty.

Some digits https://habr.com/ru/post/243285/
This can become a problem when we need to fast view a lot of files.

Probably, the parameter for setting the method will be the optimal solution.

**Comment by corvus1** [comment]
> We use `0` over 2000 years, looks like mammoth shit :)

You know exactly what I meant. :-D Even the person who wrote it in 1997 didn't think it was acceptable for downsampling, if you look at the comment in the code. 

> But example with... `rmSimpleStretch` :)))

Yes, that's exactly what we shouldn't do. That integer scaling only works one way and only in fixed integer steps. It has its uses, but this is not one of them.

> This can become a problem when we need to fast view a lot of files.

Yes, when we're making a ton thumbnails, perhaps, but not when we're flipping through images in Viewer. It'll only have to scale 1 image every time the user decides to look at another picture.

**Comment by alexx2000** [comment]
Viewer uses TImage component which scaling image by draw with scaling using widgetset capabilities.

**Comment by corvus1** [comment]
@alexx2000 I feel a bit stupid, but I poked around a little and I can't even locate roughly where it's happening in the source code. Little help? 
I was going to go take a look at the references, to see if something can be easily adjusted to change the scaling method.

**Comment by alexx2000** [comment]
https://github.com/fpc/Lazarus/blob/316d8c1de0f9e3f69abff3429085f5ed992c40e9/lcl/interfaces/qt5/qtwinapi.inc#L6819

**Comment by corvus1** [comment]
I meant in Double Commander sources. 
Especially since:
> Sizing is done according to the stretching mode currently set in the
  destination device context.

**Comment by alexx2000** [comment]
DC itself does not have image painting code. It uses [TImage](https://github.com/fpc/Lazarus/blob/316d8c1de0f9e3f69abff3429085f5ed992c40e9/lcl/extctrls.pp#L586) control which do all work.

**Comment by corvus1** [comment]
Yeah, that's what I imagined, but there is a call somewhere to that TQtWidgetSet.StretchMaskBlt function that draws the image, right? I'll go grep some more. 

And it says it draws according to the setting on the destination, so supposedly that's what we have to change. Right?

**Comment by corvus1** [comment]
I just don't quite get that.

```
corvus@desktop ~/soft/doublecmd $ grep StretchMaskBlt * -ir
grep: doublecmd.dbg: binary file matches
grep: units/x86_64-linux-gtk2/interfaces.o: binary file matches
```

**Comment by alexx2000** [comment]
TImage does not call it directly, you can find calling chain if execute DC from Lazarus under debugger with breakpoint at [TCustomImage.Paint](https://github.com/fpc/Lazarus/blob/316d8c1de0f9e3f69abff3429085f5ed992c40e9/lcl/include/customimage.inc#L275).

If you build GTK2 version then called GTK2 [StretchMaskBlt](https://github.com/fpc/Lazarus/blob/316d8c1de0f9e3f69abff3429085f5ed992c40e9/lcl/interfaces/gtk2/gtk2winapi.inc#L9688)
